### PR TITLE
add docs for `Base.return_types`

### PR DESCRIPTION
**Pull request by sjkelly**
Adds a simple docstring for this.

### CODE REVIEW

**Comment by petvana** [comment]
I'd propose to update docs for `Base.methods` as well and document the tight relationship between them. (Maybe an example with shorter result?)

**Comment by MasonProtter** [comment]
See also https://github.com/JuliaLang/julia/pull/44340

**Comment by aviatesk** [comment]
> See also #44340
> 
> I think this PR is good to draw from because it lays out explicitly what sorts of situations it is actually acceptable to use this sort of functionality.

Since `Base.return_types` is not recognized by the compiler (whereas `Core.Compiler.return_type` is so) and is just a reflection utility, it does not have the problems of `Core.Compiler.return_type`.

File: base/reflection.jl [hunk 1]
```diff
@@ -1500,6 +1500,34 @@ function code_ircode_by_type(
...
+julia> Base.return_types(+, Tuple{Any, Any})
+187-element Vector{Any}:
+ Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8}
+ Union{Float16, Float32, Float64}
+ BigInt
+ BigInt
+ BigFloat
+ BigFloat
+ BigFloat
+ Any
```
Comment (aviatesk) [inline base/reflection.jl:1524]:
```suggestion
julia> methods(sum, (Union{Vector{Int},UnitRange{Int}},))
# 2 methods for generic function "sum" from Base:
 [1] sum(r::AbstractRange{<:Real})
     @ range.jl:1396
 [2] sum(a::AbstractArray; dims, kw...)
     @ reducedim.jl:996
     
julia> Base.return_types(sum, (Union{Vector{Int},UnitRange{Int}},))
2-element Vector{Any}:
 Int64 # the result of inference on sum(r::AbstractRange{<:Real})
 Int64 # the result of inference on sum(a::AbstractArray; dims, kw...)
```
Probably this is more instructive?

File: base/reflection.jl [hunk 2]
```diff
@@ -1500,6 +1500,34 @@ function code_ircode_by_type(
     return asts
 end
 
+
+"""
+    return_types(f::Function, types::DataType=default_tt(f); world::UInt=get_world_counter(), interp::NativeInterpreter=Core.Compiler.NativeInterpreter(world))
+
+Return a list of possible return types for a given function `f` and argument types `types`.
+
```
[the remark this answers is not in the document]
Comment (sjkelly) [reply base/reflection.jl]: Thanks!

File: base/reflection.jl (hunk 1 above)
Comment (sjkelly) [reply base/reflection.jl]: Agreed, I was not clear of the symmetries with `methods` until I filed this PR

**Review by aviatesk — APPROVED** [review]

**Pull request merged**